### PR TITLE
Extract PathWatcher

### LINKED ISSUE #2678: Extract ServiceWatcher into a multi-purpose PathWatcher
Given the work done on #2676 we could extract a serviceWatcher as it could be logic agnostic.

- where would it live? (which package/module)
- any preferred name? (directoryWatcher? pathWatcher?)

### PR DESCRIPTION
**Pull request by DavidRG13**
Closes #2678 

Extract PathWatcher

### CODE REVIEW

File: core/cas-server-core-services/src/main/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDao.java [hunk 1]
```diff
@@ -48,12 +51,16 @@
         LOG_SERVICE_DUPLICATE.accept(s2);
         return s1;
     };
+    private static final int INTERVAL = 0;
```
Comment (mmoayyed) [inline core/cas-server-core-services/src/main/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDao.java]: What is the significance of this value here?

File: core/cas-server-core-services/src/main/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDao.java [hunk 2]
```diff
@@ -48,12 +51,16 @@
...
 
     /**
      * The Service registry directory.
      */
     protected Path serviceRegistryDirectory;
 
+    // PLEASE REVIEW: make sense to this publisher to live in a where ApplicationEventPublisher lives?? and just reference to it
+    private Consumer<AbstractCasEvent> publish = this::publishEvent;
```
Comment (mmoayyed) [inline core/cas-server-core-services/src/main/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDao.java]: I am not sure I follow your comment here. Could you explain please. Are you saying the publishEvent() call should turn into a separate EventPublisher component?

File: core/cas-server-core-services/src/main/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDao.java [hunk 3]
```diff
@@ -135,14 +142,14 @@ private void initializeRegistry(final Path configDirectory,
...
-                        publishEvent(new CasRegisteredServicesRefreshEvent(this));
+                        createServiceRefreshEvent.andThen(publish);
                     } else {
                         LOGGER.debug("Service [{}] loaded from [{}] is identical to the existing entry. Entry may have already been saved "
                                 + "in the event processing pipeline", newService.getId(), file.getName());
                     }
                 }
             };
-            this.serviceRegistryConfigWatcher = new ServiceRegistryConfigWatcher(serviceRegistryDirectory, onCreate, onModify, onDelete);
+            this.serviceRegistryConfigWatcher = new PathWatcher(serviceRegistryDirectory, onCreate, onModify, onDelete, INTERVAL);
```
Comment (mmoayyed) [inline core/cas-server-core-services/src/main/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDao.java]: If this actually needs to be 0 to limit monitoring intervals, how about an overloaded constructor that removes the need for the passing of that value?

File: core/cas-server-core-util/src/main/java/org/apereo/cas/util/function/ComposableFunction.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Represents a function that accepts one argument and produces a result and
+ * could be with a Consumer of the resulting type.
+ *
+ * PLEASE REMOVE THIS IF IT'S ADDED TO THE JDK
```
Comment (mmoayyed) [inline core/cas-server-core-util/src/main/java/org/apereo/cas/util/function/ComposableFunction.java]: Thanks for the note. Should this turn into a TODO comment?

File: core/cas-server-core-util/src/main/java/org/apereo/cas/util/function/ComposableSupplier.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package org.apereo.cas.util.function;
+
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+/**
+ * Represents a supplier of results that could be chained with a Consumer.
+ *
+ * PLEASE REMOVE THIS IF IT'S ADDED TO THE JDK
```
Comment (mmoayyed) [inline core/cas-server-core-util/src/main/java/org/apereo/cas/util/function/ComposableSupplier.java]: Thanks for the note. Should this turn into a TODO comment?

File: support/cas-server-support-configuration/src/main/java/org/apereo/cas/config/CasConfigurationSupportUtilitiesConfiguration.java [hunk 6]
```diff
@@ -45,7 +57,10 @@
         @Autowired
         @Qualifier("configurationPropertiesEnvironmentManager")
         private CasConfigurationPropertiesEnvironmentManager configurationPropertiesEnvironmentManager;
-        
+
+        // PLEASE REVIEW: make sense to this publisher to live in a where ApplicationEventPublisher lives?? and just reference to it
+        private Consumer<AbstractCasEvent> publish = event -> eventPublisher.publishEvent(event);
```
Comment (mmoayyed) [inline support/cas-server-support-configuration/src/main/java/org/apereo/cas/config/CasConfigurationSupportUtilitiesConfiguration.java]: See comment above please.

File: core/cas-server-core-services/src/main/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDao.java (hunk 2 above)
Comment (DavidRG13) [inline core/cas-server-core-services/src/main/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDao.java]: I mean that by this approach there is no need to pass ```applicationEventPublisher``` all around. Just use the publish consumer that expect an ```AbstractCasEvent``` or even better a ```Supplier<? extends AbstractCasEvent>```.

If this makes sense, this publish consumer should live somewhere accessible from everywhere

File: core/cas-server-core-util/src/main/java/org/apereo/cas/util/function/ComposableFunction.java (hunk 4 above)
Comment (DavidRG13) [inline core/cas-server-core-util/src/main/java/org/apereo/cas/util/function/ComposableFunction.java]: How ToDos comments are created in javadoc? I can't find anything

File: core/cas-server-core-services/src/main/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDao.java (hunk 2 above)
Comment (mmoayyed) [inline core/cas-server-core-services/src/main/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDao.java]: Yup, that's all good. Make it happen please :)

File: core/cas-server-core-util/src/main/java/org/apereo/cas/util/function/ComposableFunction.java (hunk 4 above)
Comment (mmoayyed) [inline core/cas-server-core-util/src/main/java/org/apereo/cas/util/function/ComposableFunction.java]: Not a javadoc thing; more of an IDE+Checkstyle warning thing. I'd rather be annoyed by INFO/WARN messages of some static analysis tool than to discover things need to be maintained by reading javadocs :)

File: core/cas-server-core-services/src/main/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDao.java (hunk 2 above)
Comment (DavidRG13) [inline core/cas-server-core-services/src/main/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDao.java]: From my point of view, it should be done in a separate PR, do you agree?

**Comment by DavidRG13** [comment]
How could we bypass the ToDo comment error?

File: core/cas-server-core-services/src/main/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDao.java (hunk 2 above)
Comment (mmoayyed) [inline core/cas-server-core-services/src/main/java/org/apereo/cas/services/AbstractResourceBasedServiceRegistryDao.java]: Yes, definitely.

**Comment by mmoayyed** [comment]
See http://checkstyle.sourceforge.net/config_misc.html#TodoComment

Add this check into the checkstyle rules, or edit it if it already exists and set it to issue warnings instead of errors.

**Comment by DavidRG13** [comment]
Do you know how to make Codacy ignore it?

**Comment by mmoayyed** [comment]
I removed and restarted the build. Thanks for following up.
